pkg/controller/volume/expand/cache/volume_resize_map.go

```diff
@@ -0,0 +1,303 @@
+	defer resizeMap.Unlock()
+
+	pvcUniqueName := types.UniquePvcName(pvc.UID)
+
+	if _, ok := resizeMap.pvcrs[pvcUniqueName]; !ok {
+		pvcSize := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+		pvSize := pv.Spec.Capacity[v1.ResourceStorage]
+		// If PV size is less than pvc size. The reason here we compare with pv size
+		// rather than pvc.Status.Size is because if a resize is already in progress
+		// then pv size would reflect new value and if pv size is more than pvc
+		// size then no resize will be necessary.
+		if pvSize.Cmp(pvcSize) < 0 {
+			pvcRequest := &PvcWithResizeRequest{
+				PVC:          pvc,
+				CurrentSize:  pvc.Status.Capacity[v1.ResourceStorage],
+				ExpectedSize: pvcSize,
+				VolumeSpec:   volume.NewSpecFromPersistentVolume(pv, false),
+			}
+			resizeMap.pvcrs[pvcUniqueName] = pvcRequest
+		}
+	}
+}
+
+// Return Pvcrs that require resize
+func (resizeMap *volumeResizeMap) GetPVCsWithResizeRequest() []*PvcWithResizeRequest {
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+
+	pvcrs := []*PvcWithResizeRequest{}
+	for _, pvcr := range resizeMap.pvcrs {
+		pvcrs = append(pvcrs, pvcr)
+	}
+	// Empty out pvcrs map, we will add back failed resize requests later
+	resizeMap.pvcrs = map[types.UniquePvcName]*PvcWithResizeRequest{}
+	return pvcrs
+}
+
+// DeletePVC removes given pvc object from list of pvcs that needs resizing.
+// deleting a pvc in this map doesn't affect operations that are already inflight.
+func (resizeMap *volumeResizeMap) DeletePVC(pvc *v1.PersistentVolumeClaim) {
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+	pvcUniqueName := types.UniquePvcName(pvc.UID)
+	glog.V(5).Infof("Removing PVC %v from resize map", pvcUniqueName)
+	delete(resizeMap.pvcrs, pvcUniqueName)
+}
+
+func (resizeMap *volumeResizeMap) MarkAsResized(pvcr *PvcWithResizeRequest) error {
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+
+	// If any part of this update fails, resize will be attempted again
+	err := resizeMap.updatePvSize(pvcr, pvcr.ExpectedSize)
+	if err != nil {
+		glog.V(4).Infof("Error updating PV spec capacity for volume %q with : %v", pvcr.QualifiedName(), err)
+		resizeMap.addBackResizeRequest(pvcr)
+		return err
+	}
+
+	emptyCondition := []v1.PersistentVolumeClaimCondition{}
+
+	err = resizeMap.updatePvcCapacityAndConditions(pvcr, pvcr.ExpectedSize, emptyCondition)
+	if err != nil {
+		glog.V(4).Infof("Error updating PV spec capacity for volume %q with : %v", pvcr.QualifiedName(), err)
+		resizeMap.addBackResizeRequest(pvcr)
+		return err
+	}
+	return nil
+}
+
+func (resizeMap *volumeResizeMap) MarkResizeFailed(pvcr *PvcWithResizeRequest, reason string) error {
+	resizeMap.Lock()
+	defer resizeMap.Unlock()
+
+	// This needs to be done atomically somehow so as these operations succeed or fail together.
+	// If any part of this update fails, resize has to be tried again
+	emptyCondition := []v1.PersistentVolumeClaimCondition{}
+
+	_, err := util.UpdatePVCCondition(pvcr.PVC, emptyCondition, resizeMap.kubeClient)
```

Comment:
IMO we should keep any ResizeStarted condition there, the controller is still trying to resize the PVC.

Reply:
We do set it back when we actually attempt resize in `sync_resize_map`. Should be enough may be?

Reply:
I'd leave the condition there. Constant adding + removing the condition looks weird to me. The condition should tell user that resizing is in progress on the PVC and has not finished yet. And it has not finished, the controller is just taking a break for a while.

Reply:
done